### ISSUE TITLE
Add CLI tool for listing patches available from installed plugins

### ISSUE THREAD

**Issue by kdmccormick** [openedx/wg-developer-experience#50]
## Context

Originally proposed in https://github.com/overhangio/2u-tutor-adoption/issues/5 was an in-code documentation system for patches accompanied by a CLI tool like this:
```
$ tutor config patches list
PATCH                           PLUGIN        DESCRIPTION
openedx-common-settings         core          Python to be added both LMS and
                                              CMS common settings modules
local-docker-compose-services   core          Additional services to run in local
                                              and dev mode, using docker-compose.yml syntax
credentials-certificate-footer  credentials   HTML to add to footer of course
                                              certificate (hypothetical)
...etc
```
We decided that the implementation and maintenance of this would not justify its worth.

Instead, we will document patches manually (as described in https://github.com/overhangio/2u-tutor-adoption/issues/5). Then, we could make a command-line tool just to list the patches and the locations that they will render (relative to the Tutor root), like this:
```
$ tutor config patches list
PATCH                           LOCATIONS
openedx-common-settings         env/apps/openedx/settings/cms/production.py
                                env/apps/openedx/settings/lms/production.py
local-docker-compose-services   env/dev/docker-compose.yml
credentials-certificate-footer  env/plugins/credentials/apps/credentials/html/certs/footer.html
...etc
```
## Blockers

Plugin API v1: #32 

## Acceptance

* Propose an CLI syntax and output format on the Tutor Forums.
* Implement it.

**Comment by MaferMazu** [comment]
I can take this issue if you agree, @kdmccormick :D

**Comment by kdmccormick** [comment]
All yours!

Heads up @regisb that @MaferMazu will be working on this.

**Comment by MaferMazu** [comment]
Hi, I want to ask about the implementation of this feature.

At first moment I thought it interesting to extract the info for this command from the [patches.rst](https://github.com/overhangio/tutor/blob/df0e26c58eba48dd8eaf7eadec5775f617ce0dc9/docs/reference/patches.rst), and I found two approaches:
First, the more accessible, but I don't think it is elegant, read the doc line by line and extract the patches and the associated files.
Second, try to parse the RST file, and with that structure, try to extract the info. For this approach, I need to create a Directive for patch and access to the information with a tree or something like an XML structure.

Another approach I had was like have another doc with a simple structure that provides the patch names and files, but we would maintain two docs of patch documentation.

What do you think @kdmccormick @rgraber @Carlos-Muniz ? Which is the best approach?

Another question I have is about the column output. Do you know a better way to do it than [ljust()](https://docs.python.org/3/library/stdtypes.html?highlight=ljust#str.ljust) or [{: >20}.format()](https://docs.python.org/2.7/library/string.html#format-specification-mini-language) ?

I would appreciate any feedback 😊

**Comment by regisb** [comment]
If I were to implement this feature I would not try to parse the patches.rst file at all. Instead, I would examine the values in `hooks.Filters.ENV_PATCHES.callbacks`. In particular, the `callback.context` attribute will tell you the name of the app it comes from. 

Does that make sense? This introspection is fairly low-level in the hooks API, and it's not something that is commonly done by most developers (well I hope not!). Start by implementing the command, have a look at the values above and it should work :crossed_fingers:

**Comment by kdmccormick** [comment]
From Slack:
> @MaferMazu : 
> The issue #50 is to list all patches used by all installed plugins, right?
> For example, if I have two plugins, one using openedx-common-settings and caddyfile, and the other using openedx-common-settings and openedx-common-i18n-settings, the output of the command would be only the three patches and the location of each one.
> I understand that the form to do it is analyzing the callbacks.context of [ENV_PATCHES](https://github.com/overhangio/tutor/blob/master/tutor/hooks/consts.py#L230).
> Am I on the same page? Please let me know if I am not.

**Comment by kdmccormick** [comment]
What do you think @regisb ? Personally, I cannot think of any way to do this other than recursively scanning each directory in `ENV_TEMPLATE_ROOTS` and looking for instances of `patch("...")`.

**Comment by MaferMazu** [comment]
I was trying to iterate over ENV_PATCHES, and I need to populate the Filter. So I was thinking about a pre-load, like reading the config with `config = tutor_config.load_minimal(context.root)` and then loading the plugins like [this](https://github.com/overhangio/tutor/blob/master/tutor/commands/plugins.py#L47) to obtain the patches somehow.

The other way, I think, is the idea of Kyle, is doing something with the `ENV_TEMPLATE_ROOTS,` maybe using [this](https://github.com/overhangio/tutor/blob/master/tutor/env.py#L152).

Am I on the right path?

@kdmccormick @regisb

**Comment by kdmccormick** [comment]
@regisb your approach sounds good, but to zoom in on the patches-within-patches problem:

> For instance, if a plugin implements the "caddyfile" patch, and in this patch there is a {{ patch("custompatch") }} statement, then we will not be able to associate the "custompatch" to any given template.
>
> This is a relatively minor detail, but let's keep it in mind.

The `{{ patch ("custompatch") }}` will still be rendered, right? So we should still be able to print it out:
```
```

**Comment by MaferMazu** [comment]
I was thinking of adding `self.current_template = template_name` in the [render template function](https://github.com/overhangio/tutor/blob/master/tutor/env.py#L167), and in the patch function creates a file to store the patch name and the template that patches edit: but only store the patches that are modified by a plugin. Then when we execute the `tutor config patches list` command only read that file created. What do you think? Do you have a better approach? @kdmccormick @regisb @Carlos-Muniz @rgraber 


About the @kdmccormick comment, I haven't used patches-within-patches; if you have an example, I appreciate it for better understanding. ✨

**Comment by kdmccormick** [comment]
Sorry for the slow response @MaferMazu.

Saving the list to a file is an interesting idea, but I do not prefer it because it would persist state between commands. When we rely on state, there is extra room for errors. For example, if someone installed a plugin and then ran `tutor config patches list` before running `tutor config save`, then the file would be missing or outdated.

Instead, I recommend following Régis's advice above, which is to subclass the `Renderer`  class and override its `render_template` and `patch` functions so that they `print(...)` each patch occurrence. Then, this new subclass could be used to implement the `tutor config patches list` command. Does that approach make sense?

**Comment by MaferMazu** [comment]
Yes, it makes sense. So I'm going to start working on that. I only have a question, and it's about where patches include patches; if you have an example, that would be helpful.

**Comment by kdmccormick** [comment]
> where patches include patches

@MaferMazu , I could have sworn that some plugins had patches that included other patches, but to be honest I cannot find any examples. Sorry about that! When we extract LMS & CMS into a plugin, I expect that we will have plenty of examples, but for now I'll just have to invent one:

Consider [tutor-mfe's implement of the **caddyfile** patch](https://github.com/overhangio/tutor-mfe/blob/v15.0.4/tutormfe/patches/caddyfile), which adds routing for the `.apps.` subdomain for MFEs (`learning.apps.example.com`, `profile.apps.example.com`, etc.). Now, imagine that we wanted to allow *other plugins* to add custom routes to `.apps.`. We might change the patch to this:
```
{{ MFE_HOST }}{$default_site_port} {
    {{ patch("caddyfile-mfe") | indent(2) }}  # <------------------------- This line was added

    respond / 204
    request_body {
        max_size 2MB
    }

    reverse_proxy /api/mfe_config/v1* lms:8000 {
        # We need to specify the host header, otherwise it will be rejected with 400
        # from the lms.
        header_up Host {{ LMS_HOST }}
    }
    import proxy "mfe:8002"
}
```
Now, imagine we create a *tutor-mfe-custom* plugin. That plugin could implement our new **caddyfile-mfe** patch as this:
```
respond /custom-path 200 {
  body "This is a custom response!"
  close
}
```
then `http://apps.example.com/custom-path` would return the custom response.

To summarize:
* The caddyfile template in Tutor core exposes a **caddyfile** patch.
* *tutor-mfe* implements the **caddyfile** patch, and exposes its own **caddyfile-mfe** patch.
* *tutor-mfe-custom* implements the **caddyfile-mfe** patch ("a patch within a patch").

At the moment, this is an edge case. If handling it is causing problems in your implementation, I would say go ahead and ignore the patch-within-a-patch edge case. If the edge case becomes common in the future, we can always come back and add support for it 😄

**Comment by MaferMazu** [comment]
Kyle, thanks for your answer.
I started a draft PR here https://github.com/overhangio/tutor/pull/798 for this issue; I will refactor it and then open the PR for your review.
I think I forgot to list only the patches we use with our plugins. We want to list only the patches we are using, right? cc @kdmccormick

**Comment by kdmccormick** [comment]
@MaferMazu Great!

We want to list all patches, whether they are used or not, and whether they are from the core or from plugins. That way, when somebody is a developing a new plugin, they can use this tool in order to see all patches that are available for them to use.

**Comment by MaferMazu** [comment]
@kdmccormick, the PR for this is ready for review. Maybe it's a good idea to link this issue with that PR and set this card in review status.
PR: https://github.com/overhangio/tutor/pull/798

**Comment by kdmccormick** [comment]
I don't have permission to link the PR to this issue unfortunately, but I've marked it as In Review.

**Comment by MaferMazu** [comment]
@kdmccormick I think we can close this issue; the PR https://github.com/overhangio/tutor/pull/798 was merged.

**Comment by kdmccormick** [comment]
🎉
